src/lib.rs

```diff
@@ -1110,6 +1135,52 @@ impl<'a, T> Iterator for Drain<'a, T> {
     }
 }
 
+/// This `struct` is created by [`HashBag::difference`].
+/// See its documentation for more.
+pub struct Difference<'a, T, S = RandomState> {
+    /// An iterator over `self` items
+    items: IterInner<'a, T>,
+
+    /// The bag with entries we DO NOT want to return
+    other: &'a HashBag<T, S>,
+
+    /// For `size_hint()`
+    upper_bound: usize,
+}
+
+impl<'a, T: fmt::Debug> fmt::Debug for Difference<'a, T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.debug_struct("Difference")
+            .field("items", &self.items)
+            .field("other", &self.other)
+            .finish()
+    }
+}
+
+impl<'a, T, S> Iterator for Difference<'a, T, S>
+where
+    T: Eq + Hash,
+    S: BuildHasher,
+{
+    type Item = (&'a T, usize);
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let (t, n) = self.items.next()?;
+            let other_n = self.other.contains(t);
+            if other_n < *n {
+                return Some((t, *n - other_n));
+            }
+        }
```

Comment:
I think this needs to also reduce the upper bound for each iteration. It's not _incorrect_ not to do so, but since the iterator can be inspected at any time, we should tighten the bound as we can.

Reply:
Fixed in the latest commit using the Dr-Emann trick.